db/model/generator.js

```diff
@@ -475,22 +570,29 @@ module.exports = function generator(seq, dataTypes) {
    * currentCollector field and expects the caller to save later.
    */
   Generator.prototype.assignToCollector = function () {
-    const possibleCollectors = this.possibleCollectors;
-    let newColl;
-    if (this.isActive && possibleCollectors && possibleCollectors.length) {
-      possibleCollectors.sort((c1, c2) => c1.name > c2.name);
-      newColl = possibleCollectors.find((c) => c.isRunning() && c.isAlive());
-    }
-
-    logAssignment(this.name, this.currentCollector, newColl);
-
-    // We could use setCurrentCollector, but that would result in database saves
-    // that would be unnecessary and complicate our logic in the db hooks.
-    // Instead, we set the foreign key (collectorId) from collector model.
-    // We also mock the currentCollector on the instance to avoid needing
-    // a database reload.
-    this.collectorId = newColl ? newColl.id : null;
-    this.currentCollector = newColl || null;
+    return Promise.resolve()
+    .then(() => {
+      const possibleCollectors = this.possibleCollectors;
+      if (this.isActive && possibleCollectors && possibleCollectors.length) {
+        return getMostAvailableCollector(possibleCollectors)
+        .then((mostAvailCollector) => {
+          logAssignment(this.name, this.currentCollector, mostAvailCollector);
+
+          // We could use setCurrentCollector, but that would result in database
+          // saves that would be unnecessary and complicate our logic in the db
+          // hooks. Instead, we set the foreign key (collectorId) from collector
+          // model. We also mock the currentCollector on the instance to avoid
+          // needing a database reload.
+          this.collectorId = mostAvailCollector ? mostAvailCollector.id : null;
+          this.currentCollector = mostAvailCollector || null;
+          return Promise.resolve();
+        });
+      }
+
+      this.collectorId = null;
+      this.currentCollector = null;
+      return Promise.resolve();
```

Comment:
does it make testing any easier or more transparent if this fn returns the collector it assigned to?

Reply:
Hmm - thats good point - what do you think @jgraff2